### PR TITLE
changefeedccl: add roachtest testing msk serverless

### LINKED ISSUE #124432: changefeedccl: support aws msk serverless
CDC already supports AWS MSK (Amazon's managed kafka service) via the kafka sink. This issue is to add any syntactic sugar to support connections to AWS MSK serverless directly, as well as to test changefeeds that use AWS MSK serverless as a sink.

Jira issue: CRDB-38891

Epic CRDB-31660

### PR DESCRIPTION
**Pull request by asg0451**
Add a roachtest to test the kafka sink against msk serverless.

Resolves: #124432

Release note: None

### CODE REVIEW

File: pkg/cmd/roachtest/tests/cdc.go [hunk 1]
```diff
@@ -1913,6 +1921,56 @@ func registerCDC(r registry.Registry) {
...
+			defer stopFeeds(db)
+
+			tdb := sqlutils.MakeSQLRunner(db)
+			tdb.Exec(t, `CREATE TABLE auth_test_table (a INT PRIMARY KEY)`)
+
+			testCerts, err := makeTestCerts("0.0.0.0", "aws") // ?
+			require.NoError(t, err)
+			feeds := map[string]string{
+				"scram": fmt.Sprintf("kafka://%s?tls_enabled=true&ca_cert=%s&sasl_enabled=true&sasl_user=scram512&sasl_password=scram512-secret&sasl_mechanism=SCRAM-SHA-512", brokers.scram, testCerts.CACertBase64()),
+				"iam":   fmt.Sprintf("kafka://%s?tls_enabled=true&sasl_enabled=true&sasl_mechanism=AWS_MSK_IAM&sasl_aws_region=us-east-2&sasl_aws_iam_role_arn=arn:aws:iam::541263489771:role/miles-msk-testing&sasl_aws_iam_session_name=miles", brokers.iam),
```
Comment (asg0451) [inline pkg/cmd/roachtest/tests/cdc.go:1955]: TODO: make a new iam role with the same permissions but a different name and use that

File: pkg/cmd/roachtest/tests/cdc.go [hunk 2]
```diff
@@ -2610,6 +2668,10 @@ func (k kafkaManager) binDir() string {
 	return filepath.Join(k.basePath(), k.confluentInstallBase(), "bin")
 }
 
+func (k kafkaManager) libDir() string {
```
Comment (asg0451) [inline pkg/cmd/roachtest/tests/cdc.go:2671]: TODO: rm

**Review by andyyang890 — COMMENTED** [review]
Did an initial pass, looks pretty good!

> Reviewed 6 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-:-O1niAh01GXDOv3RY7dG:bnm5kus)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @renatolabs, @rharding6373, and @vidit-bhat)

___
*[`-- commits` line 6 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1njWpX2zK9SibChW8t):*
Did you mean to tag https://github.com/cockroachdb/cockroach/issues/124432 instead?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1955 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/126954#gh-1672997886:-O1nk8Bo9F9-mUKSrEh6:b6ajmer) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3754210ccadd7c7232fea960ebbcb78f33939cc/pkg/cmd/roachtest/tests/cdc.go#L1955)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

TODO: make a new iam role with the same permissions but a different name and use that
</blockquote></details>

I presume this is `arn:aws:iam::541263489771:role/roachprod-msk-full-access` found below?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1926 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1n4LCF4H8L_ceebTfM:-O1n4LCF4H8L_ceebTfN:b-c1ee0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L1926)):*
> ```Go
> 	r.Add(registry.TestSpec{
> 		Name:             "cdc/kafka-auth-msk",
> 		Owner:            `cdc`,
> ```

nit: s/\`cdc\`/registry.OwnerCDC
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1934 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nb3R59H5MVO5RdV8M:-O1nb3R59H5MVO5RdV8N:b-i8gd0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L1934)):*
> ```Go
> 		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
> 			c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings(), c.All())
> 			mkm := NewMskManager(ctx, t, msktypes.ClusterTypeServerless)
> ```

nit: s/mkm/mskManager
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3460 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nbRNE5TwqT9V-M4Lk:-O1nbRNE5TwqT9V-M4Ll:bbpmwj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3460)):*
> ```Go
> 	t           test.Test
> 	mskClient   *msk.Client
> 	of          func(*msk.Options)
> ```

nit: s/of/mskOptsFn

imo struct fields should err on the side of having longer names since they have much longer lifetimes compared to a local variable
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3469 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nbAskENUi9jZPa83a:-O1nbAskENUi9jZPa83b:b-gmibaz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3469)):*
> ```Go
> const awsRegion = "us-east-2"
> 
> func NewMskManager(ctx context.Context, t test.Test, clusterType msktypes.ClusterType) *mskManager {
> ```

nit: s/Msk/MSK
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3497 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1ndeOM3h3X4GSstBd5:-O1ndeOM3h3X4GSstBd6:b-qnam83) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3497)):*
> ```Go
> 		m.t.Fatalf("provisioned msk clusters are not yet supported in this utility")
> 	case msktypes.ClusterTypeServerless:
> 		var region roachprodaws.AWSRegion
> ```

Should there be some defensive code after the for loop to handle if the region is not found? e.g. fataling the test with a useful error message
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nf5Sg5VEQJBzO0y6Y:-O1nf5Sg5VEQJBzO0y6Z:bjrh4wg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3533)):*
> ```Go
> 	}
> 
> 	_, err := m.mskClient.PutClusterPolicy(ctx, &msk.PutClusterPolicyInput{
> ```

Do you mind adding a comment explaining the purpose of this policy?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1ngkCc3grWnsNhGaF-:-O1ngkCc3grWnsNhGaF0:bs81l6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3664)):*
> ```Go
> 	withCTN := option.WithNodes(createTopicNode)
> 
> 	require.NoError(m.t, c.RunE(ctx, withCTN, "mkdir", "-p", "/tmp/create-msk-topic"))
> ```

nit: maybe `/tmp/create-msk-topic` should be extracted into a const?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3667 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1ngxOHE1nsSqM4TBRc:-O1ngxOHE1nsSqM4TBRd:b-cel3nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3667)):*
> ```Go
> 	require.NoError(m.t, c.PutString(ctx, createMskTopicMain, "/tmp/create-msk-topic/main.go", 0700, createTopicNode))
> 	require.NoError(m.t, c.PutString(ctx, setupMskTopicScript, "/tmp/create-msk-topic/run.sh", 0700, createTopicNode))
> 	require.NoError(m.t, c.RunE(ctx, withCTN, "/tmp/create-msk-topic/run.sh", m.connectInfo.iamBroker, topic))
> ```

It seems kind of wasteful to have to repeat this process if we want to create multiple topics. Could the copying and building instead happen once in a `SetUp` function?
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3670 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nfLUg7UorzWG6yuiV:-O1nfLUg7UorzWG6yuiW:b8r4e6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3670)):*
> ```Go
> }
> 
> func (m *mskManager) Teardown() {
> ```

nit: s/Teardown/TearDown
___
*[`pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1n30aN1lkTybxt5r9m:-O1n30aOClCg7xcyhXd5:bo1u0md) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt#L1)):*
> ```help
> package main
> ```

Any reason to not make this a normal `.go` file? Also, maybe it should go in a separate `testdata` or `scripts` subdirectory? 

One idea: `scripts/create-msk-topic/main.go`
___
*[`pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1niY7A5r5mCOsZ1wKP:-O1niY7A5r5mCOsZ1wKQ:bgy2ss9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt#L1)):*
> ```help
> package main
> ```

Could you also add a comment at the very top of this file explaining this file is going to be copied to a roachprod node during the test?

**Review by andyyang890 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @asg0451, @renatolabs, @rharding6373, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3489 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nlOUZ1KsJphfSBA9H:-O1nlOUZ1KsJphfSBA9I:b-gxxv4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3489)):*
> ```Go
> }
> 
> func (m *mskManager) MakeCluster(ctx context.Context) {
> ```

Any reason in particular that you exported the methods but not the data type? In any case, exported methods should have comments. Ditto for the constructor and all the other methods below.

**Review by asg0451 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @renatolabs, @rharding6373, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1926 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1n4LCF4H8L_ceebTfM:-O1qsRZN9mqkZI75C2Px:b-2803ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L1926)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: s/\`cdc\`/registry.OwnerCDC
</blockquote></details>

fixed but many of these tests don't use that constant so maybe we should fix that at some point
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3489 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nlOUZ1KsJphfSBA9H:-O1qtMFgEFg4jNWTAjku:b-39yws4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3489)):*
> exported the methods

it's a compromise -- i want to give the struct a definite api that roachtests should use, without actually exporting the struct which is unnecessary at this time. 

> exported methods should have comments

while generally good practice I don't think it's necessary here, since the type itself isnt exported and this is roachtest only code anyway
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3667 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1ngxOHE1nsSqM4TBRc:-O1rKEbZ0NLXZNiodHll:b-a8wrmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3667)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

It seems kind of wasteful to have to repeat this process if we want to create multiple topics. Could the copying and building instead happen once in a `SetUp` function?
</blockquote></details>

I think that can be punted until we want to set up multiple topics. All cdc kafka roachtests use one topic currently i think.
___
*[`pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1n30aN1lkTybxt5r9m:-O1rKeG40bbbJYOmsKEv:bsmrtyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt#L1)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Any reason to not make this a normal `.go` file? Also, maybe it should go in a separate `testdata` or `scripts` subdirectory? 

One idea: `scripts/create-msk-topic/main.go`
</blockquote></details>

making this an actual .go file can mess with go modules / bazel. we could consider integrating it with bazel, but i think this is the least weird of the available options

- call bazel build in the roachtest :-1: 
- somehow plumb the binary into the roachtest like cockroach (how do we set this up on ci??)
- set it up with its own go.mod, go.sum, main.go, and tar it up to the roachprod instance and build it there. but bazel/go will think it's part of our code tree when it isnt.

**Review by rharding6373 — COMMENTED** [review]
Thanks for the work on this! Have a couple comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-:-O1rf3A32Ki95DferDAR:b-mx1x3q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @renatolabs, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 3489 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1nlOUZ1KsJphfSBA9H:-O1rXKpp08LT8PGBdvrq:b-vsbi0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/cdc.go#L3489)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

> exported the methods

it's a compromise -- i want to give the struct a definite api that roachtests should use, without actually exporting the struct which is unnecessary at this time. 

> exported methods should have comments

while generally good practice I don't think it's necessary here, since the type itself isnt exported and this is roachtest only code anyway
</blockquote></details>

Please add method comments.
___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1regcL-i1fT_Pa1Ssx:-O1regcM1sMeGwmK71Fg:b-2bp80e) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf4966c091b8ffb4fb59587007a3fe96867bff42/pkg/cmd/roachtest/tests/cdc.go#L1959)):*
> ```Go
> 				}
> 			}
> 			if brokers.scramConnectURI != "" { // Not available with serverless.
> ```

I thought this roachtest only uses serverless? Do we need to include this conditional code path?
___
*[`pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1n30aN1lkTybxt5r9m:-O1rdm_l9qMR3DDKvoIS:b-dis692) ([raw file](https://github.com/cockroachdb/cockroach/blob/ec502033c9bf893c45f993d763ea0aae3d4643af/pkg/cmd/roachtest/tests/create-msk-topic-main-go.txt#L1)):*
<details><summary><i>Previously, asg0451 (Miles Frankel) wrote…</i></summary><blockquote>

making this an actual .go file can mess with go modules / bazel. we could consider integrating it with bazel, but i think this is the least weird of the available options

- call bazel build in the roachtest :-1: 
- somehow plumb the binary into the roachtest like cockroach (how do we set this up on ci??)
- set it up with its own go.mod, go.sum, main.go, and tar it up to the roachprod instance and build it there. but bazel/go will think it's part of our code tree when it isnt.
</blockquote></details>

I think this methodology is ok (not really that different from other scripts we copy during tests), but I would also move it to a subdirectory.

**Review by asg0451 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @renatolabs, @rharding6373, and @vidit-bhat)

___
*[`pkg/cmd/roachtest/tests/cdc.go` line 1959 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-O1regcL-i1fT_Pa1Ssx:-O1rqJDA9kHRj6XbIuT4:b-qyg5kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf4966c091b8ffb4fb59587007a3fe96867bff42/pkg/cmd/roachtest/tests/cdc.go#L1959)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

I thought this roachtest only uses serverless? Do we need to include this conditional code path?
</blockquote></details>

we're going to want to add msk provisioned as well no? just not in this PR.

**Review by rharding6373 — APPROVED** [review]
:lgtm: Thanks for bringing this over the finish line!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/126954#-:-O7AFSmH3mgx1eJT8yFu:bnfukh6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @andyyang890, @renatolabs, and @vidit-bhat)

**Comment by asg0451** [comment]
bors r=rharding6373

**Pull request merged**